### ISSUE TITLE
Introduce an abstraction for client's "implicit grab"

### ISSUE THREAD

**Issue by vberger** [Smithay/smithay#106]
The wayland protocol relies on several places on what is called "implicit grabs". They represent temporary exclusivity given to clients to perform some kinds of actions in response to input events.

The general process is that events initiating an implicit grab provide a "serial" (`u32`) to the client, and this client is required to give back the serial when attempting to do an action requiring an implicit grab. There is a single implicit grab per seat. A client request using a grab serial that is no longer valid will in general be ignored, but is not a protocol error.

As such, we need to introduce some abstraction to manage these implicit grabs and their serial. It would keep track of the current implicit grab for each seat, and allow to check if a client currently ows a grab. So I'd imagine an API like:

```rust
impl GrabManager {
    fn new_grab(&self, seat: &Seat) -> u32 { ... }
    fn is_current_grab(&self, serial: u32) -> bool { ... }
}
```

with `GrabManager` being some wrapper around an `Arc<...>`, allowing it to be shared accros various parts of the compositor.

This is likely necessary for #49 and #28 .

**Comment by vberger** [comment]
Actually, the grabs need to be checked per-seat sometimes, so these should probably rather be new methods on `Seat`, but the source of serials should be common to all seats, so that no two grabs can have the same serial.

**Comment by vberger** [comment]
I have gathered some understand regarding grabs in wayland. As always with the wayland protocol, this is actually more subtle than it looks.

So, wayland has a notion of "grab" of an input device. Some protocol action can trigger an *explicit* grab of an input device, altering its behavior for the duration of the grab.

Some examples:

- Starting a drag'n'drop action will grab the pointer, changing its behavior for as long as the mouse button is pressed, to act on the drag'n'drop mechanism.
- An xdg_shell popup can grab the cursor and keyboard until it is removed or the user clicks on a surface belonging to an other client.

The protocol also has an implicit notion of grab, which represents the default grabbing mechanism when no explicit grab is in process.

Example:

- When an user clicks a surface then moves the cursor outside of it while maintaining the button pressed, the surface will keep keyboard focus until the button is released.

**Comment by vberger** [comment]
solved by #109
